sys/net/gnrc/link_layer/gomach/gomach.c

```diff
@@ -0,0 +1,2244 @@
+            /* If we haven't reach the maximum t2u limit, try again. Set quit_current_cycle flag
+             * to true such that we will release the current neighbor pointer.  */
+            gnrc_gomach_set_quit_cycle(netif, true);
+            netif->mac.tx.t2u_state = GNRC_GOMACH_T2U_END;
+        }
+
+        gnrc_gomach_set_update(netif, true);
+        return;
+    }
+
+    /* If we didn't catch the RX-start event, go to send the next preamble. */
+    if ((gnrc_gomach_timeout_is_expired(netif, GNRC_GOMACH_TIMEOUT_PREAMBLE)) ||
+        gnrc_gomach_get_max_pream_interv(netif)) {
+        netif->mac.tx.t2u_state = GNRC_GOMACH_T2U_PREAMBLE_PREPARE;
+        gnrc_gomach_set_update(netif, true);
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_PREAMBLE);
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_MAX_PREAM_INTERVAL);
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_WAIT_RX_END);
+    }
+}
+
+static void gomach_t2u_send_data(gnrc_netif_t *netif)
+{
+    /* If we are retrying to send the data, reload its original MAC sequence. */
+    if (netif->mac.tx.no_ack_counter > 0) {
+        netdev_ieee802154_t *device_state = (netdev_ieee802154_t *)netif->dev;
+        device_state->seq = netif->mac.tx.tx_seq;
+    }
+
+    /* Here, we send the data to the receiver. */
+    int res = gnrc_gomach_send_data(netif, NETOPT_ENABLE);
+    if (res < 0) {
+        LOG_ERROR("ERROR: [GOMACH] t2u data sending error: %d.\n", res);
+
+        /* If res is < 0, the data packet will not be released in send().
+         * so need to release the data here. */
+        if (netif->mac.tx.packet != NULL) {
+            gnrc_pktbuf_release(netif->mac.tx.packet);
+            netif->mac.tx.packet = NULL;
+        }
+
+        netif->mac.tx.t2u_state = GNRC_GOMACH_T2U_END;
+        gnrc_gomach_set_update(netif, true);
+        return;
+    }
+
+    gnrc_gomach_set_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR, GNRC_GOMACH_NO_TX_ISR_US);
+
+    netif->mac.tx.t2u_state = GNRC_GOMACH_T2U_WAIT_DATA_TX;
+    gnrc_gomach_set_update(netif, false);
+}
+
+static void gomach_t2u_wait_tx_feedback(gnrc_netif_t *netif)
+{
+    if (gnrc_gomach_timeout_is_expired(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR)) {
+        /* No TX-ISR, go to sleep. */
+        netif->mac.tx.t2u_retry_counter++;
+
+        netif->mac.tx.no_ack_counter = GNRC_GOMACH_REPHASELOCK_THRESHOLD;
+        netdev_ieee802154_t *device_state = (netdev_ieee802154_t *)netif->dev;
+        netif->mac.tx.tx_seq = device_state->seq - 1;
+
+        gnrc_gomach_set_quit_cycle(netif, true);
+        netif->mac.tx.t2u_state = GNRC_GOMACH_T2U_END;
+
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR);
+        gnrc_gomach_set_update(netif, true);
+        return;
+    }
+
+    if (gnrc_gomach_get_tx_finish(netif)) {
+        gnrc_gomach_clear_timeout(netif, GNRC_GOMACH_TIMEOUT_NO_TX_ISR);
+
+        if (gnrc_netif_get_tx_feedback(netif) == TX_FEEDBACK_SUCCESS) {
```

Comment:
Is it be possible to make each block of the `if { } else { }` a function tat is called there?

Reply:
Addressed~